core/trino-main/src/main/java/io/trino/sql/planner/EqualityInference.java

```diff
@@ -273,20 +266,18 @@ public static EqualityInference newInstance(Metadata metadata, Collection<Expres
 
     private Expression rewrite(Expression expression, Predicate<Symbol> symbolScope, boolean allowFullReplacement)
     {
-        Set<Expression> subExpressions = SubExpressionExtractor.extract(expression);
-        if (!allowFullReplacement) {
-            subExpressions = subExpressions.stream()
-                    .filter(e -> !e.equals(expression))
-                    .collect(Collectors.toSet());
-        }
-
         ImmutableMap.Builder<Expression, Expression> expressionRemap = ImmutableMap.builder();
-        for (Expression subExpression : subExpressions) {
-            Expression canonical = getScopedCanonical(subExpression, symbolScope);
-            if (canonical != null) {
-                expressionRemap.put(subExpression, canonical);
-            }
-        }
+        SubExpressionExtractor.extract(expression)
+                .distinct()
```

Comment:
you can remove `distinct` if you use `Map<Expression, Expression> expressionRemap = new HashMap()` instead
Consider adding checks that previous mapping is identical

Reply:
Ok, removed distinct, used hashmap + putIfAbsent.